### PR TITLE
 feat: Implement JavaScript Coverage

### PR DESCRIPTION
**Pull request by aslushnikov**
This patch introduces a new `page.coverage` namespace with two methods:
- `page.coverage.startJSCoverage` to initiate JavaScript coverage
  recording
- `page.coverage.stopJSCoverage` to stop JavaScript coverage and get
  results

### CODE REVIEW

**Comment by JoelEinbinder** [comment]
I am getting inconsistent errors like `Protocol error (Debugger.getScriptSource): No script for id: 127 undefined`

My coverage code looks like:
```js
const files = []
afterAll(SX(async function () {
  server.close();
  browser.close();
  fs.writeFile(path.join(__dirname, 'coverage.json'), JSON.stringify(files), function() {});
}));
beforeEach(SX(async function () {
  await page.coverage.startJSCoverage();
  await page.goto('http://localhost:3000/test.html')
  files.push(await page.coverage.stopJSCoverage());
}));
```

File: docs/api.md [hunk 1]
```diff
@@ -2143,6 +2151,39 @@ Identifies what kind of target this is. Can be `"page"`, `"service_worker"`, or
...
+### class: Coverage
+
+#### coverage.startJSCoverage()
+- returns: <[Promise]> Promise that resolves when coverage is started
+
+#### coverage.stopJSCoverage()
+- returns: <[Promise]<[Array]<[Object]>>> Promise that resolves to the array of coverage reports for all non-anonymous scripts
+  - `url` <[string]> Script URL
+  - `text` <[string]> Script content
+  - `ranges` <[Array]<[Object]>> Parts of the script that were executed
```
Comment (JoelEinbinder) [inline docs/api.md]: Are these ranges in order? Can ranges overlap?

File: docs/api.md [hunk 2]
```diff
@@ -2143,6 +2151,39 @@ Identifies what kind of target this is. Can be `"page"`, `"service_worker"`, or
...
+await page.goto('https://example.com');
+const coverage = await page.stopJSCoverage();
+let totalBytes = 0;
+let usedBytes = 0;
+for (const entry of coverage) {
+  totalBytes += entry.text.length;
+  for (const range of entry.ranges)
+    usedBytes += range.endOffset - range.startOffset - 1;
+}
+console.log(`Coverage is ${usedBytes/totalBytes*100}%`);
```
Comment (vsemozhetbyt) [inline docs/api.md]: Should the operators be spaced?

Comment (aslushnikov) [reply docs/api.md]: Done.

File: docs/api.md (hunk 1 above)
Comment (aslushnikov) [reply docs/api.md]: Ranges are sorted and non-overlapping. Added this to the description.

File: docs/api.md [hunk 3]
```diff
@@ -2143,6 +2151,39 @@ Identifies what kind of target this is. Can be `"page"`, `"service_worker"`, or
 #### target.url()
 - returns: <[string]>
 
+### class: Coverage
+
+#### coverage.startJSCoverage()
```
Comment (JoelEinbinder) [inline docs/api.md]: Add note about coverage not surviving navigations. I might try starting coverage at the start of my tests, ending at the end of my tests, and reporting sections that are not used. We use this pattern ourselves.

File: lib/Coverage.js [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+  async startJSCoverage() {
+    console.assert(!this._jsCoverage, 'JSCoverage is already enabled');
+    this._jsCoverage = true;
+    this._scripts.clear();
+    this._jsEventListeners = [
+      helper.addEventListener(this._client, 'Debugger.scriptParsed', this._onScriptParsed.bind(this)),
+      helper.addEventListener(this._client, 'Runtime.executionContextsCleared', this._onExecutionContextsCleared.bind(this))
+    ];
+    await Promise.all([
+      this._client.send('Profiler.enable'),
```
Comment (JoelEinbinder) [inline lib/Coverage.js]: Is the order of these methods important? Are we relying on them being synchronous on the backend?

File: lib/Coverage.js [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+      this._client.send('Profiler.startPreciseCoverage', {callCount: false, detailed: true}),
+      this._client.send('Debugger.enable'),
+      this._client.send('Debugger.setSkipAllPauses', {skip: true})
+    ]);
+  }
+
+  _onScriptParsed(event) {
```
Comment (JoelEinbinder) [inline lib/Coverage.js]: Can we jsdoc this event?

File: lib/Coverage.js (hunk 4 above)
Comment (aslushnikov) [reply lib/Coverage.js]: yes: we should enable profiler before starting coverage, and similarly we're enabling debugger before skipping pauses.

File: lib/Coverage.js (hunk 5 above)
Comment (aslushnikov) [reply lib/Coverage.js]: jsdoc'ed it like everywhere else (though it's rather useless jsdoc)

File: test/test.js [hunk 6]
```diff
@@ -3397,6 +3397,69 @@ describe('Page', function() {
...
+
+  describe('JSCoverage', function() {
+    it('should work', async function({page, server}) {
+      await page.coverage.startJSCoverage();
+      await page.goto(server.PREFIX + '/jscoverage/simple.html');
+      const coverage = await page.coverage.stopJSCoverage();
+      expect(coverage.length).toBe(1);
+      expect(coverage[0].url).toContain('/jscoverage/simple.html');
+      expect(coverage[0].ranges).toEqual([
+        { startOffset: 0, endOffset: 17 },
```
Comment (JoelEinbinder) [inline test/test.js]: Considering that you have just created a file format, I'd feel safer with some large golden-file tests.

File: lib/Coverage.js (hunk 4 above)
Comment (JoelEinbinder) [reply lib/Coverage.js]: Will this conflict if we use profiler for something else later?

File: lib/Coverage.js [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+    const scriptIdToSource = new Map();
+    await Promise.all(profileResponse.result.map(async entry => {
+      // Ignore profile entries that don't have scripts.
+      if (!scripts.has(entry.scriptId))
+        return;
+      try {
+        const response = await this._client.send('Debugger.getScriptSource', {scriptId: entry.scriptId});
+        scriptIdToSource.set(entry.scriptId, response.scriptSource);
+      } catch (e) {
+        // This means there's no script source for the script. This might
```
Comment (JoelEinbinder) [inline lib/Coverage.js]: debugerror?

**Review by JoelEinbinder — APPROVED** [review]

File: test/test.js (hunk 6 above)
Comment (aslushnikov) [reply test/test.js]: Added a golden file, though not a very large one.

**Review by JoelEinbinder — APPROVED** [review]

**Pull request merged**